src/core/ext/filters/client_channel/xds/xds_bootstrap.cc

```diff
@@ -39,129 +39,84 @@ std::unique_ptr<XdsBootstrap> XdsBootstrap::ReadFromFile(grpc_error** error) {
-      }
-      seen_xds_servers = true;
-      grpc_error* parse_error = ParseXdsServerList(child);
-      if (parse_error != GRPC_ERROR_NONE) error_list.push_back(parse_error);
-    } else if (strcmp(child->key, "node") == 0) {
-      if (child->type != GRPC_JSON_OBJECT) {
-        error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
-            "\"node\" field is not an object"));
-      }
-      if (seen_node) {
-        error_list.push_back(
-            GRPC_ERROR_CREATE_FROM_STATIC_STRING("duplicate \"node\" field"));
-      }
-      seen_node = true;
-      grpc_error* parse_error = ParseNode(child);
-      if (parse_error != GRPC_ERROR_NONE) error_list.push_back(parse_error);
-    }
-  }
-  if (!seen_xds_servers) {
+  auto it = json.mutable_object()->find("xds_servers");
+  if (it == json.mutable_object()->end()) {
     error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
         "\"xds_servers\" field not present"));
+  } else if (it->second.type() != Json::Type::ARRAY) {
+    error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
+        "\"xds_servers\" field is not an array"));
+  } else {
+    grpc_error* parse_error = ParseXdsServerList(&it->second);
+    if (parse_error != GRPC_ERROR_NONE) error_list.push_back(parse_error);
+  }
+  it = json.mutable_object()->find("node");
+  if (it != json.mutable_object()->end()) {
+    if (it->second.type() != Json::Type::OBJECT) {
+      error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
+          "\"node\" field is not an object"));
+    } else {
+      grpc_error* parse_error = ParseNode(&it->second);
+      if (parse_error != GRPC_ERROR_NONE) error_list.push_back(parse_error);
+    }
   }
   *error = GRPC_ERROR_CREATE_FROM_VECTOR("errors parsing xds bootstrap file",
                                          &error_list);
 }
 
-XdsBootstrap::~XdsBootstrap() {
-  grpc_json_destroy(tree_);
-  grpc_slice_unref_internal(contents_);
-}
-
-grpc_error* XdsBootstrap::ParseXdsServerList(grpc_json* json) {
+grpc_error* XdsBootstrap::ParseXdsServerList(Json* json) {
   InlinedVector<grpc_error*, 1> error_list;
   size_t idx = 0;
-  for (grpc_json *child = json->child; child != nullptr;
-       child = child->next, ++idx) {
-    if (child->key != nullptr) {
-      char* msg;
-      gpr_asprintf(&msg, "array element %" PRIuPTR " key is not null", idx);
-      error_list.push_back(GRPC_ERROR_CREATE_FROM_COPIED_STRING(msg));
-    }
-    if (child->type != GRPC_JSON_OBJECT) {
+  for (Json& child : *json->mutable_array()) {
+    if (child.type() != Json::Type::OBJECT) {
       char* msg;
       gpr_asprintf(&msg, "array element %" PRIuPTR " is not an object", idx);
       error_list.push_back(GRPC_ERROR_CREATE_FROM_COPIED_STRING(msg));
     } else {
-      grpc_error* parse_error = ParseXdsServer(child, idx);
+      grpc_error* parse_error = ParseXdsServer(&child, idx);
       if (parse_error != GRPC_ERROR_NONE) error_list.push_back(parse_error);
     }
```

Comment:
Should be an `idx++` here?

Reply:
Good catch!  Sent you #21866 to fix this.